### PR TITLE
DROOLS-5491 - Integrate PMML with Kogito

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @mariofusco @jiripetrlik

see https://issues.redhat.com/browse/DROOLS-5491

This PR
1) implement standalone PMML evaluation inside kogito-quarkus
2) implement DMN-PMML integration inside kogito-quarkus

It needs https://github.com/kiegroup/drools/pull/2994

### CODE REVIEW

File: kogito-quarkus-extension/deployment/src/main/java/org/kie/kogito/quarkus/deployment/KogitoAssetsProcessor.java [hunk 1]
```diff
@@ -390,54 +415,66 @@ private ApplicationGenerator createApplicationGenerator(AppPaths appPaths, Combi
...
+        ApplicationGenerator appGen = new ApplicationGenerator(appPackageName,
+                                                               new File(appPaths.getFirstProjectPath().toFile(),
+                                                                        "target"))
                 .withDependencyInjection(new CDIDependencyInjectionAnnotator())
                 .withAddons(addonsConfig)
                 .withGeneratorContext(context);
 
         addProcessGenerator(appPaths, addonsConfig, appGen);
         addRuleGenerator(appPaths, appGen, addonsConfig);
+        addPredictionGenerator(appPaths, appGen, addonsConfig);
```
Comment (danielezonca) [inline kogito-quarkus-extension/deployment/src/main/java/org/kie/kogito/quarkus/deployment/KogitoAssetsProcessor.java:427]: As far as I can see this and the implementation of this method are the only real change in the whole file.
Can you please revert the formatting? I think should increase the limit of characters per line

File: api/kogito-api/pom.xml [hunk 2]
```diff
@@ -50,6 +50,35 @@
...
+        </exclusion>
+        <exclusion>
+          <groupId>org.kie</groupId>
+          <artifactId>kie-api</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
+    <dependency>
+      <groupId>org.kie</groupId>
+      <artifactId>kie-pmml-commons</artifactId>
```
Comment (jiripetrlik) [inline api/kogito-api/pom.xml:69]: Is there any reason to add kie-pmml-commons to kogito API? I have some concerns to add this to kogito-api bacause this should be only a public API which does not change often. With this we are exposing also kie-pmml-commons as public API. I would suggest to refactor this somehow in the way that api itself does not depend on kie-pmml-commons.

File: api/kogito-api/src/main/java/org/kie/kogito/StaticConfig.java [hunk 3]
```diff
@@ -24,20 +25,23 @@
     private final Addons addons;
     private final ProcessConfig processConfig;
     private final RuleConfig ruleConfig;
+    private final PredictionConfig predictionConfig;
     private final DecisionConfig decisionConfig;
 
-    public StaticConfig(Addons addons, ProcessConfig processConfig, RuleConfig ruleConfig, DecisionConfig decisionConfig) {
+    public StaticConfig(Addons addons, ProcessConfig processConfig, RuleConfig ruleConfig,  DecisionConfig decisionConfig, PredictionConfig predictionConfig) {
```
Comment (jiripetrlik) [inline api/kogito-api/src/main/java/org/kie/kogito/StaticConfig.java:31]: Please remove one space before "DecisionConfig".

File: drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/config/DefaultPredictionEventListenerConfig.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+package org.kie.kogito.pmml.config;
+
+import org.kie.kogito.prediction.PredictionEventListenerConfig;
+
+public class DefaultPredictionEventListenerConfig implements PredictionEventListenerConfig {
+
+    public DefaultPredictionEventListenerConfig() {
+    }
+
+
```
Comment (jiripetrlik) [inline drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/config/DefaultPredictionEventListenerConfig.java:25]: Remove unnecessary new lines.

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        final String predictionRuleMapperSource =
+                PredictionRuleMapperGenerator.getPredictionRuleMapperSource(fullRuleName);
+        return new org.drools.modelcompiler.builder.GeneratedFile( org.drools.modelcompiler.builder.GeneratedFile.Type.CLASS, predictionRuleMapperPath, predictionRuleMapperSource);
+    }
+
+    private void storeFile(GeneratedFile.Type type, String path, String source) {
+        generatedFiles.add(new GeneratedFile(type, path, log(source).getBytes(StandardCharsets.UTF_8)));
+    }
+
+
```
Comment (jiripetrlik) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:287]: Remove new lines.

File: api/kogito-api/src/main/java/org/kie/kogito/prediction/PredictionRuleMapper.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.kogito.prediction;
+
+public interface PredictionRuleMapper {
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply api/kogito-api/src/main/java/org/kie/kogito/prediction/PredictionRuleMapper.java:18]: I would prefer to keep it here.
Spreading code here and there makes maintenance pretty hard.

File: api/kogito-api/src/main/java/org/kie/kogito/prediction/PredictionConfig.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.kogito.prediction;
+
+public interface PredictionConfig {
+
+    PredictionEventListenerConfig predictionEventListeners();
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply api/kogito-api/src/main/java/org/kie/kogito/prediction/PredictionConfig.java:20]: Preparing for EventListener implementation

see https://issues.redhat.com/browse/DROOLS-5193

File: kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/util/MojoUtil.java [hunk 8]
```diff
@@ -52,14 +53,8 @@
...
-            final File file = artifact.getFile();
-            if (file != null && file.isFile()) {
-                urls.add(file.toURI().toURL());
-                final KieModuleModel depModel = getDependencyKieModel(file);
-                if (kmoduleDeps != null && depModel != null) {
-                    final ReleaseId releaseId = new ReleaseIdImpl(artifact.getGroupId(), artifact.getArtifactId(), artifact.getVersion());
-                    kmoduleDeps.add(new ZipKieModule(releaseId, depModel, file));
-                }
+            if (artifact.getType().equals("jar")) {
+                populateURLsFromJarArtifact(urls, artifact, kmoduleDeps);
```
Comment (danielezonca) [inline kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/util/MojoUtil.java:57]: What is the reason of this change?

File: drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/PMMLKogito.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        }
+    }
+
+    public static PMML4Result evaluate(PMMLRuntime pmmlRuntime, String modelName, Map<String, Object> pmmlContext) {
+        final PMMLRequestData pmmlRequestData = getPMMLRequestData(modelName, pmmlContext);
+        return pmmlRuntime.evaluate(modelName, new PMMLContextImpl(pmmlRequestData));
+    }
+
+    private static PMMLRequestData getPMMLRequestData(String modelName, Map<String, Object> parameters) {
+        String correlationId = "CORRELATION_ID";
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/PMMLKogito.java:78]: @danielezonca 
It is not a constant. It is a left-over from legacy implementation and I still have to find a clean way to use/generate it.

File: drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/PMMLRuntimeBuilder.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+import org.kie.pmml.evaluator.api.container.PMMLPackage;
+import org.kie.pmml.evaluator.api.executor.PMMLRuntime;
+import org.kie.pmml.evaluator.assembler.container.PMMLPackageImpl;
+import org.kie.pmml.evaluator.core.executor.PMMLModelEvaluatorFinderImpl;
+import org.kie.pmml.evaluator.core.service.PMMLRuntimeImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.kie.pmml.evaluator.assembler.service.PMMLAssemblerService.getFactoryClassNamePackageName;
+import static org.kie.pmml.evaluator.assembler.service.PMMLLoaderService.getKiePMMLModelsLoadedFromResource;
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply drools/kogito-pmml/src/main/java/org/kie/kogito/pmml/PMMLRuntimeBuilder.java]: This would make sense only if/when there would a PMML module implemented outside any "container" - so that common code (e.g. PMMLRuntimeBuilder) could be referred to by both containers (Drools/Kogito) without having one of them depending on the other.

File: api/kogito-api/src/main/java/org/kie/kogito/prediction/PredictionModel.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package org.kie.kogito.prediction;
+
+import java.util.Map;
+
+import org.kie.api.pmml.PMML4Result;
+import org.kie.dmn.api.core.DMNContext;
+import org.kie.dmn.api.core.DMNResult;
+import org.kie.pmml.commons.model.KiePMMLModel;
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply api/kogito-api/src/main/java/org/kie/kogito/prediction/PredictionModel.java:23]: Yes - not sure to remember why it ended here

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+            return Collections.emptyList();
+        }
+        ModelBuilderImpl<KogitoPackageSources> modelBuilder = new ModelBuilderImpl<>( KogitoPackageSources::dumpSources,
+                new KnowledgeBuilderConfigurationImpl(getClass().getClassLoader()), new ReleaseIdImpl("dummy:dummy:0.0.0"), true, false );
+        CompositeKnowledgeBuilder batch = modelBuilder.batch();
+        for (PMMLResource resource : resources) {
+            List<KiePMMLModel> kiepmmlModels = resource.getKiePmmlModels();
+            for (KiePMMLModel model : kiepmmlModels) {
+                if (model.getName() == null || model.getName().isEmpty()) {
+                    throw new RuntimeException("Model name should not be empty");
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PredictionCodegen.java:206]: Model name is automatically generated (when missing) now.

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PMMLResource.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.nio.file.Path;
+import java.util.List;
+
+import org.kie.pmml.commons.model.KiePMMLModel;
+
+public class PMMLResource {
+    private final List<KiePMMLModel> kiePmmlModels;
+    private final Path path;
+    private final String modelPath;
```
[the remark this answers is not in the document]
Comment (gitgabrio) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/prediction/PMMLResource.java:26]: They are different
See PredictionCodegen#110 
`pmmlResources.addAll(parsePredictions(jarPath, resources));`

File: kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/util/MojoUtil.java (hunk 8 above)
Comment (gitgabrio) [reply kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/util/MojoUtil.java:57]: If you look the original code - and the invoked method - they both expected a file to be a "jar" file, which is not verified anywhere. An artifact may contain a "pom" file, and in such situation the invoked method inside MojoUtil.getDependencyKieModel(final File jar)
`ZipFile zipFile = new ZipFile(jar)`
throws an Excpetion

File: drools/drools-core-static/src/main/java/org/drools/statics/StaticServiceRegistry.java [hunk 14]
```diff
@@ -55,11 +57,15 @@ private void wireServices() {
...
         serviceMap.put(KieAssemblers.class, new StaticKieAssemblers());
+        serviceMap.put(KieRuntimes.class, SimpleInstanceCreator.instance("org.kie.internal.services.KieRuntimesImpl"));
+
 
         registerService("org.drools.compiler.kie.builder.impl.InternalKieModuleProvider", "org.drools.modelcompiler.CanonicalKieModuleProvider", true);
         registerService("org.drools.compiler.compiler.DecisionTableProvider", "org.drools.decisiontable.DecisionTableProviderImpl", false);
 
         constructorMap.put("TimerService", SimpleInstanceCreator.constructor("org.kie.kogito.timer.impl.JDKTimerService"));
+
+        registerKieRuntimeService("org.kie.pmml.evaluator.api.executor.PMMLRuntime", "org.kie.pmml.evaluator.core.service.PMMLRuntimeService");
```
Comment (rsynek) [inline drools/drools-core-static/src/main/java/org/drools/statics/StaticServiceRegistry.java:68]: @gitgabrio @danielezonca I have stumbled upon this when trying to upgrade kogito to the latest snapshot in optaplanner. 
Does it mean that if I declare a dependency on org.kie.kogito:drools-core, which depends on drools-core-static, I have to declare also a runtime dependency on org.kie:kie-pmml-evaluator-core (or rather org.kie.kogito:kogito-pmml)?

Comment (gitgabrio) [reply drools/drools-core-static/src/main/java/org/drools/statics/StaticServiceRegistry.java:68]: @rsynek 
See https://issues.redhat.com/browse/KOGITO-3290
